core/controllers/ingest.py

```diff
@@ -361,3 +364,34 @@ def remove_unreferenced_organisations():
             db.session.delete(organisation)
 
     db.session.commit()
+
+
+def with_transaction(
+    function: callable, max_retries: int, sleep_duration: float, error_type: Type[Exception], **kwargs: dict
+) -> None:
+    """Execute a transaction with retries for a specified error type.
+
+    This function wraps the provided 'function' in a database transaction and allows for retries
+    in case of a specified error type. If an error of the specified type occurs during the
+    transaction, the operation is retried up to a specified maximum number of times with a
+    delay between retries.
+
+    :param function: The function to be executed within the database transaction.
+    :param max_retries: The maximum number of retries in case of an error.
+    :param sleep_duration: The duration to sleep between retry attempts.
+    :param error_type: The type of error to catch and retry.
+    :param kwargs: Additional keyword arguments to pass to the 'function'.
+    :return: None
+    """
+    for retry in range(max_retries):
+        try:
+            with db.session.begin():
+                function(**kwargs)
+            break
+        except error_type as transaction_error:
+            if retry < max_retries:
+                max_retries -= 1
+                time.sleep(sleep_duration)
```

Comment:
Do we currently log anything other than errors? If so we could log the retry here?